cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,219 @@
+	MinCursor = Cursor{}
+	// MaxCursor is the largest possible cursor
+	MaxCursor = Cursor{
+		Ledger: math.MaxUint32,
+		Tx:     math.MaxUint32,
+		Op:     math.MaxUint32,
+		Event:  math.MaxUint32,
+	}
+)
+
+type event struct {
+	contents xdr.ContractEvent
+	cursor   Cursor
+}
+
+func cmp(a, b uint32) int {
+	if a < b {
+		return -1
+	}
+	if a > b {
+		return 1
+	}
+	return 0
+}
+
+// Cmp compares two cursors.
+// 0 is returned if the c is equal to other.
+// 1 is returned if c is greater than other.
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			return cmp(c.Event, other.Event)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock            sync.RWMutex
+	events          []event
+	length          int
+	start           int
+	retentionWindow uint32
+}
+
+// NewMemoryStore creates a new MemoryStore populated by the given ledgers.
+func NewMemoryStore(passPhrase string, ledgers []xdr.LedgerCloseMeta, retentionWindow uint32) (*MemoryStore, error) {
+	var events []event
+	for _, ledger := range ledgers {
+		reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(passPhrase, ledger)
+		if err != nil {
+			return nil, err
+		}
+		ledgerEvents, err := readEvents(reader)
+		if err != nil {
+			return nil, err
+		}
+		events = append(events, ledgerEvents...)
+	}
+
+	return &MemoryStore{
+		events:          events,
+		retentionWindow: retentionWindow,
+		length:          len(events),
+	}, nil
+}
+
+// Scan applies f on all the events occurring in the range of [start, end).
+// The events are processed in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (m *MemoryStore) Scan(start, end Cursor, f func(Cursor, xdr.ContractEvent) bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	i := sort.Search(m.length, func(i int) bool {
+		index := (m.start + i) % m.length
+		entry := m.events[index]
+		return entry.cursor.Cmp(start) < 0
+	}) + 1
+
+	for ; i < m.length; i++ {
+		index := (m.start + i) % m.length
+		entry := m.events[index]
+		if entry.cursor.Cmp(end) >= 0 {
+			break
+		}
+		if !f(entry.cursor, entry.contents) {
+			break
+		}
+	}
+}
+
+// IngestEvents adds new events from the given ledger into the store.
+// As a side effect, events which fall outside the retention window are
+// removed from the store.
+func (m *MemoryStore) IngestEvents(txReader *ingest.LedgerTransactionReader) error {
+	events, err := readEvents(txReader)
+	if err != nil {
+		return err
+	}
+	if len(events) == 0 {
+		return nil
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.trim(txReader.GetSequence())
+	m.maybeResize(len(events))
+	m.append(events)
+	return nil
+}
+
+func readEvents(txReader *ingest.LedgerTransactionReader) ([]event, error) {
```

Comment:
this function requires to be executed under write lock; please add a comment about that.
( the implementation is fine; but we need to add a comment so it would be clear )